File: julia/2_quadratic_equation.jl
# quadratic_equation(a,b,c): a function that computes the real roots of a
# given quadratic equation a*X^2+b*X+c=0.

# using the quadratic formula:
# If there is one real root, it returns a list with one element
# If there are no real roots, it prints a message and returns an empty vector.

function quadratic_equation(a, b, c)
    # Handle degenerate case when a=0
    if a == 0
        if b == 0
            error("Not a valid equation: a and b cannot both be 0.")
        else
            # Linear equation: b*x + c = 0 => x = -c/b
            return [-c / b]
        end
    end
    # Calculate the discriminant
    discriminant = b^2 - 4 * a * c
    if discriminant < 0
        println("No real roots.")
        return []
    elseif discriminant == 0
        # One real root
        x = -b / (2*a)
        return [x]
    else
        # Two real roots
        x1 = (-b + sqrt(discriminant)) / (2 * a)
        x2 = (-b - sqrt(discriminant)) / (2 * a)
        return [x1, x2]
    end
end

# Example usage:
roots = quadratic_equation(1, -3, 2)  # Equation: x^2 - 3x + 2 = 0
println("Quadratic Equation Roots: ", roots)
# Quadratic Equation Roots: [2.0, 1.0]

File: python/2_quadratic_equation.py
# quadratic_equation(A,B,C): a function that computes the real roots of a
# given quadratic equation A*X^2+B*X+C=0.

import math

def quadratic_equation(a, b, c):
    """
    Compute the real roots of a*x^2 + b*x + c = 0.
    Returns a list of zero, one, or two real roots.
    """
    # Degenerate case: not actually quadratic
    if a == 0:
        if b == 0:
            return []             # no equation or infinite solutions—treat as “no real roots”
        return [-c / b]           # linear solution

    # Quadratic case
    disc = b*b - 4*a*c  # calculate the discriminant
    if disc < 0:
        return []                 # no real roots
    elif disc == 0:
        return [-b / (2*a)]      # one real (double) root
    else:
        sqrt_disc = math.sqrt(disc)
        return [(-b + sqrt_disc)/(2*a), (-b - sqrt_disc)/(2*a)]

# Example usage:
print(quadratic_equation(1, -3, 2))  # [2.0, 1.0]
print(quadratic_equation(1, 2, 1))  # [-1.0]
print(quadratic_equation(1, 0, -4))  # [2.0, -2.0]
print(quadratic_equation(0, 2, 1))  # [-0.5]
print(quadratic_equation(0, 0, 1))  # []